### PR TITLE
fix the watch cache

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 1]
```diff
@@ -277,6 +279,26 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 	if err != nil {
 		return fmt.Errorf("couldn't compute key: %v", err)
 	}
+
+	{ // hack
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go]: I'll have to rework it, just to get it going.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 2]
```diff
@@ -1125,7 +1127,8 @@ func (lw *cacherListerWatcher) List(options metav1.ListOptions) (runtime.Object,
 		Continue: options.Continue,
 	}
 
-	if err := lw.storage.List(context.TODO(), lw.resourcePrefix, storage.ListOptions{ResourceVersionMatch: options.ResourceVersionMatch, Predicate: pred}, list); err != nil {
+	ctx := genericapirequest.WithCluster(context.TODO(), genericapirequest.Cluster{Name: logicalcluster.Wildcard, Wildcard: true})
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: it allows us to bring all available data

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 3]
```diff
@@ -277,6 +279,26 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
...
+			ns = t.GetObjectMeta().GetNamespace()
+		default:
+			klog.Infof("could not get ClusterName on object: %T", event.Object)
+		}
+		if len(ns) > 0 && len(clusterName) > 0 {
+			sk := strings.Split(key, "/")
+			key = fmt.Sprintf("%v/%v/%v", strings.Join(sk[:len(sk)-2], "/"), clusterName, strings.Join(sk[len(sk)-2:], "/"))
+		} else if len(clusterName) > 0 {
+			sk := strings.Split(key, "/")
+			key = fmt.Sprintf("%v/%v/%v", strings.Join(sk[:len(sk)-1], "/"), clusterName, strings.Join(sk[len(sk)-1:], "/"))
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go]: recomputing the key under we store objects will be necessary

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 4]
```diff
@@ -277,6 +279,26 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
...
+		clusterName := ""
+		ns := ""
+		switch t := event.Object.(type) {
+		case metav1.ObjectMetaAccessor:
+			clusterName = t.GetObjectMeta().GetClusterName()
+			ns = t.GetObjectMeta().GetNamespace()
+		default:
+			klog.Infof("could not get ClusterName on object: %T", event.Object)
+		}
+		if len(ns) > 0 && len(clusterName) > 0 {
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go]: Can we get a test or at least a comment with expected inputs and outputs? This is a bit hard to read with all the splits and joins

Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go]: Also was the reason we needed to do this ... that watch cache is computing keys in some other way from the keyfunc in listers?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go]: done. The reflector used by the watch cache uses `logicalcluster. Wildcard` to discover all resources for the given GVR. During that process, the clusterName will be assigned by the storage layer. The watch cache needs to rewrite the URLs for each resource to contain the clusterName as this is how the clients will ask for resources.

**Comment by p0lyn0mial** [comment]
this PR is ready for review PTAL. Thanks!

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_cluster.go [hunk 5]
```diff
@@ -94,3 +101,10 @@ func ClusterNameFrom(ctx context.Context) (logicalcluster.Name, error) {
 	}
 	return cluster.Name, nil
 }
+
+func IsFromStorageDecorator(ctx context.Context) bool {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context_cluster.go]: Go doc, waht does this mean?

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_cluster.go [hunk 6]
```diff
@@ -29,6 +29,9 @@ type clusterKey int
 const (
 	// clusterKey is the context key for the request namespace.
 	clusterContextKey clusterKey = iota
+
+	// cacheWatermarkContextKey is the context key to indicate the key function is being used by a storage decorator
+	storageDecoratorsWatermarkContextKey clusterKey = iota
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context_cluster.go]: what does it mean? By a decorator? Can you expand a bit in the comment?

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 7]
```diff
@@ -240,7 +240,9 @@ func NoNamespaceKeyRootFunc(ctx context.Context, prefix string) string {
 	key := prefix
 	cluster, err := genericapirequest.ValidClusterFrom(ctx)
 	if err != nil {
-		klog.Errorf("invalid context cluster value: %v", err)
+		if !genericapirequest.IsFromStorageDecorator(ctx) {
+			klog.Errorf("invalid context cluster value: %v", err)
+		}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: worth to mark as KCP change?

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 8]
```diff
@@ -1421,11 +1423,15 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
...
 		}
 
+		// keyFunc is used by the storage decorators and there is no way to pass a context
+		// thus we annotate it with a watermark to indicate requests are allowed to have an empty cluster name
+		// e.g. in the case of the watch cache we don't know a cluster name upfront
+		ctx := genericapirequest.WithStorageDecoratorWatermark(genericapirequest.NewContext())
 		if isNamespaced {
-			return e.KeyFunc(genericapirequest.WithNamespace(genericapirequest.NewContext(), accessor.GetNamespace()), accessor.GetName())
+			return e.KeyFunc(genericapirequest.WithNamespace(ctx, accessor.GetNamespace()), accessor.GetName())
 		}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1426]: worth to mark as KCP change?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 9]
```diff
@@ -96,6 +99,9 @@ type Config struct {
 	Codec runtime.Codec
 
 	Clock clock.Clock
+
+	// KcpIndicators hold some KCP indicators that are used by the reflector to instruct the storage layer how to extract the cluster name
+	KcpIndicators context.Context
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: what is an indicator?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
+package cacher
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go:1]: patch_watch_cache.go is our pattern, isn't it?

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go [hunk 11]
```diff
@@ -99,6 +100,11 @@ type ConfigForResource struct {
 
 	// GroupResource is the relevant one
 	GroupResource schema.GroupResource
+
+	// The following fields hold config required/specific to KCP
+	//
+	// KcpIndicators hold some KCP indicators that are used by the reflector to instruct the storage layer how to extract the cluster name
+	KcpIndicators context.Context
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go]: what is an indicator on a context? Having a ctx here looks strange. Convince me it isn't :-)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 9 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: I use this term for metadata that are needed by the storage layer. I can rename to `KcpStorageLayerMetadata`

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go (hunk 11 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go]: sure, I can make it a struct, will be more explicit.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go]: essentially a struct that will hold some metadata for the storage layer - passed by the reflector.

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go (hunk 7 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: what do you mean by that?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 9 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: KcpExtraStorageMetadata?

Can you give examples here in the go doc ?

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go (hunk 11 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go]: but why in a context? Why not typed?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go]: no specific reason, let's make it more concrete, I will change it to a struct.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+type clusterNameAndNsGetter interface {
+	GetClusterName() string
+	GetNamespace() string
+}
+
+// recalculateWatchCacheKey adjusts the key under which the given object will be stored in the watch cache.
+//
+// resources in the db are stored without the clusterName, since the reflector used by the cache uses the logicalcluster.Wildcard
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go:24]: This is a touch hard to follow - all objects get their clusterName assigned by the storage layer, regardless of wildcard or not. Is the cache pre-calculating the key in a non-cluster-aware manner, and that's why we are updating a previous key here, as opposed to just generating it correctly in the first place?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+// because this is how clients are going to be retrieving data from the cache.
+func recalculateWatchCacheKey(existingKey string, object runtime.Object) string {
+	var clusterName, ns string
+	key := existingKey
+
+	switch t := object.(type) {
+	case metav1.ObjectMetaAccessor:
+		clusterName = t.GetObjectMeta().GetClusterName()
+		ns = t.GetObjectMeta().GetNamespace()
+	case clusterNameAndNsGetter:
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go]: What fulfills this interface and not ObjectMetaAccessor?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+		return key
+	}
+
+	// if we are dealing with a scoped resource put a clusterName before the namespace
+	// for example:
+	//
+	//  /group/resource/namespace/instance => /group/resource/clusterName/namespace/instance
+	//
+	//  /group/resource/$identity/namespace/instance => /group/resource/$identity/clusterName/namespace/instance
+	if len(ns) > 0 {
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go]: To what extent do we need to control edge case behavior here? We haven't checked that the `ns` exists in the key yet, nor that it is at the 2nd position by path separators

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 15]
```diff
@@ -1421,11 +1423,15 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
 			return "", err
 		}
 
+		// keyFunc is used by the storage decorators and there is no way to pass a context
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: What does it mean "no way to pass a context" - aren't you passing contexts in the next lines? :)

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: actually, I have just realized we should change all the layers to accept the context. That would simplify this PR and allow to reuse of some code. One sec :)

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: @stevekuznetsov pushed, ptal

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go (hunk 13 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go]: by `unstructured.Unstructured`

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go (hunk 10 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go:1]: I think it is correct, we have `store_patch.go`, here:
https://github.com/kcp-dev/kubernetes/blob/a8b2000cd5878fb3d0b8dbc74bec068c0a034a8e/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go (hunk 13 above)
Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go]: Maybe worth a comment, I am surprised that it's true

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 16]
```diff
@@ -1101,17 +1115,19 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
...
-func NewCacherListerWatcher(storage storage.Interface, resourcePrefix string, newListFunc func() runtime.Object) cache.ListerWatcher {
+func NewCacherListerWatcher(storage storage.Interface, resourcePrefix string, newListFunc func() runtime.Object, kcpExtraStorageMetadata *storagebackend.KcpStorageMetadata) cache.ListerWatcher {
 	return &cacherListerWatcher{
-		storage:        storage,
-		resourcePrefix: resourcePrefix,
-		newListFunc:    newListFunc,
+		storage:                storage,
+		resourcePrefix:         resourcePrefix,
+		newListFunc:            newListFunc,
+		kcpClusterAwareContext: createKCPClusterAwareContext(kcpExtraStorageMetadata),
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: storing a context still feels strange. Why not store `kcpExtraStorageMetadata` and call `createKCPClusterAwareContext` further down?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: storing a ctx seems easier here because I need to create the context only once and reuse it multiple times

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 17]
```diff
@@ -1140,7 +1156,15 @@ func (lw *cacherListerWatcher) Watch(options metav1.ListOptions) (watch.Interfac
 	if utilfeature.DefaultFeatureGate.Enabled(features.EfficientWatchResumption) {
 		opts.ProgressNotify = true
 	}
-	return lw.storage.WatchList(context.TODO(), lw.resourcePrefix, opts)
+	return lw.storage.WatchList(lw.createKCPClusterAwareContext(), lw.resourcePrefix, opts)
+}
+
+func (lw *cacherListerWatcher) createKCPClusterAwareContext() context.Context {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: nit: `func createKCPClusterAwareContext(meta *storagebackend.KcpStorageMetadata)`.

rule-of-thumb: functional (= non-side-effect) funcs should not be methods.

**Review by sttts — APPROVED** [review]
